### PR TITLE
model_dump tool for model inspection

### PR DESCRIPTION
**Pull request by dreiss**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56868 model_dump tool for model inspection**

Summary:
See __init__.py for a summary of the tool.
The following sections are present in this initial version
- Model Size.  Show the total model size, as well as a breakdown by
  stored files, compressed files, and zip overhead.  (I expect this
  breakdown to be a bit more useful once data.pkl is compressed.)
- Model Structure.  This is basically the output of
  `show_pickle(data.pkl)`, but as a hierarchical structure.
  Some structures cause this view to crash right now, but it can be
  improved incrementally.
- Zip Contents.  This is basically the output of `zipinfo -l`.
- Code.  This is the TorchScript code.  It's integrated with a blame
  window at the bottom, so you can click "Blame Code", then click a bit
  of code to see where it came from (based on the debug_pkl).  This
  currently doesn't render properly if debug_pkl is missing or
  incomplete.
- Extra files (JSON).  JSON dumps of each json file under /extra/, up to
  a size limit.
- Extra Pickles.  For each .pkl file in the model, we safely unpickle it
  with `show_pickle`, then render it with `pprint` and include it here
  if the size is not too large.  We aren't able to install the pprint
  hack that thw show_pickle CLI uses, so we get one-line rendering for
  custom objects, which is not very useful.  Built-in types look fine,
  though.  In particular, bytecode.pkl seems to look fine (and we
  hard-code that file to ignore the size limit).

I'm checking in the JS dependencies to avoid a network dependency at
runtime.  They were retrieved from the following URLS, then passed
through a JS minifier:
  https://unpkg.com/htm@3.0.4/dist/htm.module.js?module
  https://unpkg.com/preact@10.5.13/dist/preact.module.js?module

Test Plan:
Manually ran on a few models I had lying around.
Mostly tested in Chrome, but I also poked around in Firefox.
Unit test (100% coverage except for burn_in_info and main).
Ran on all internal production mobile models (but didn't look at the
results.)

Differential Revision: [D28020849](https://our.internmc.facebook.com/intern/diff/D28020849)

### CODE REVIEW

File: torch/utils/model_dump/__main__.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env python
+import sys
+from . import main
+
+sys.exit(main(sys.argv))
```
Comment (dhruvbird) [inline torch/utils/model_dump/__main__.py:5]: TIL: `__main__.py`
I've always seen `if __name__ == '__main__':` so far.

File: torch/utils/model_dump/code.js [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+
+const html = htm.bind(h);
+
+const BURNED_IN_MODEL_INFO = null;
+
+// https://stackoverflow.com/a/20732091
+function humanFileSize(size) {
+  if (size == 0) { return "0 B"; }
+  var i = Math.floor( Math.log(size) / Math.log(1024) );
+  return (size / Math.pow(1024, i)).toFixed(2) * 1 + ' ' + ['B', 'kB', 'MB', 'GB', 'TB'][i];
```
Comment (dhruvbird) [inline torch/utils/model_dump/code.js:12]: Q. Why `* 1`?

Comment (dhruvbird) [reply torch/utils/model_dump/code.js:12]: nvm, saw the link on SO.

File: torch/utils/model_dump/__init__.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+def hierarchical_pickle(data):
+    if isinstance(data, (bool, int, float, str, type(None))):
+        return data
+    if isinstance(data, list):
+        return [hierarchical_pickle(d) for d in data]
+    if isinstance(data, tuple):
+        return {
+            "__tuple_values__": hierarchical_pickle(list(data)),
+        }
```
Comment (dhruvbird) [inline torch/utils/model_dump/__init__.py:88]: okay - since JSON doesn't have a "tuple" representation.

File: torch/utils/model_dump/__init__.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        for zi in zf.infolist():
+            if not zi.filename.endswith(".py"):
+                continue
+            with zf.open(zi) as handle:
+                raw_code = handle.read()
+            with zf.open(zi.filename + ".debug_pkl") as handle:
+                raw_debug = handle.read()
+            debug_info = pickle.loads(raw_debug)
+            code_parts = []
+            for di, di_next in zip(debug_info, debug_info[1:]):
```
Comment (dhruvbird) [inline torch/utils/model_dump/__init__.py:204]: Q. This seems to miss the last entry in `debug_info`. Is this significant?

File: torch/utils/model_dump/__main__.py (hunk 1 above)
Comment (dreiss) [reply torch/utils/model_dump/__main__.py:5]: Yeah.  I was getting an error when trying to use "python -m torch.utils.model_dump" that told me to use this instead.

Comment (dreiss) [reply torch/utils/model_dump/__main__.py:5]: I think related to the fact that this is __init__.py instead of model_dump.py.  The distinction between a package and module in Python is mysterious to me.

File: torch/utils/model_dump/__init__.py (hunk 4 above)
Comment (dreiss) [reply torch/utils/model_dump/__init__.py:207]: That's a good point.  I think maybe I observed that there was always an empty entry at the end?  I'm not sure, though.  This could be a bug.

File: torch/utils/model_dump/__init__.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+        zip_files=zip_files,
+        interned_strings=list(interned_strings),
+        code_files=code_files,
+        model_data=model_data,
+        extra_files_jsons=extra_files_jsons,
+        extra_pickles=extra_pickles,
+    )}
+
+
+def get_inline_skeleton():
```
Comment (dhruvbird) [inline torch/utils/model_dump/__init__.py:272]: This is awesome!

File: torch/utils/model_dump/__init__.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+def burn_in_info(skeleton, info):
+    """Burn model info into the HTML skeleton.
+
+    The result will render the hard-coded model info and
+    have no external network dependencies for code or data.
+    """
+
+    return skeleton.replace(
+        "BURNED_IN_MODEL_INFO = null",
+        "BURNED_IN_MODEL_INFO = " + json.dumps(info).replace("/", "\\/"))
```
Comment (dhruvbird) [inline torch/utils/model_dump/__init__.py:298]: Q. why did you need to do this?

File: torch/utils/model_dump/__init__.py (hunk 4 above)
Comment (dreiss) [reply torch/utils/model_dump/__init__.py:207]: Okay, confirmed that I'm silently dropping data in at least one case.  Will fix.

File: torch/utils/model_dump/__init__.py (hunk 6 above)
Comment (dreiss) [reply torch/utils/model_dump/__init__.py:298]: The whole thing, or the slashes?

File: torch/utils/model_dump/code.js [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+    }
+
+    const model = this.state.model.model;
+
+    let error_msg = "";
+    if (err) {
+      error_msg = html`<h2 style="background:red">An error occurred.  Check console</h2>`;
+    }
+
+    return html`
```
Comment (dhruvbird) [inline torch/utils/model_dump/code.js:465]: TIL: backticks in javascript.
https://betterprogramming.pub/javascript-how-backticks-work-de269e0fb8ba

File: torch/utils/model_dump/code.js [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+      // TODO: Maybe show simple lists and tuples on one line.
+      return true;
+    }
+    if (data.__module_type__) {
+      return true;
+    }
+    if (data.__tensor_v2__) {
+      return false;
+    }
+    throw new Error("TODO: handle dict, etc.");
```
Comment (dhruvbird) [inline torch/utils/model_dump/code.js:139]: Great call out! Agree, these will show up pretty soon.

File: torch/utils/model_dump/code.js [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+    super();
+    this.state = { shown: false };
+
+    this.INLINE_TYPES = new Set(["boolean", "number", "string"])
+    this.IGNORED_STATE_KEYS = new Set(["training", "_is_full_backward_hook"])
+  }
+
+  click() {
+    const shown = !this.state.shown;
+    this.setState({shown: shown});
```
[the remark this answers is not in the document]
Comment (dreiss) [reply torch/utils/model_dump/code.js:115]: Copy-pasted from Hider.  BTW, I think the way Hider works is totally bad style, but I don't really know how to do it properly and I don't want to let it block me.

File: torch/utils/model_dump/code.js [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+  render({data, indent, prefix}, {shown}) {
+    const exp = this.expando(data) ? html`<span class=caret onClick=${() => this.click()} >${caret(shown)} </span>` : "";
+    const headline = this.renderHeadline(data);
+    const body = shown ? this.renderBody(indent, data) : "";
+    return html`${indent}${exp}${prefix}${headline}${body}`;
+  }
+}
+
+function ZipContentsSection({model: {zip_files}}) {
+  // TODO: Add human-readable sizes?
```
Comment (dhruvbird) [inline torch/utils/model_dump/code.js:247]: I think non-human readable is fine since one probably wants to see the actual size of the file.

File: torch/utils/model_dump/code.js [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+  // TODO: Add sorting options?
+  // TODO: Add hierarchical collapsible tree?
```
Comment (dhruvbird) [inline torch/utils/model_dump/code.js:249]: I think both of these are wishlist type of things. For a pytorch model, I'd rather have everything presented to me on one page.

Comment (dreiss) [reply torch/utils/model_dump/code.js:249]: No reason we couldn't have both. :)

**Review by dhruvbird — APPROVED** [review]

File: torch/utils/model_dump/code.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (dhruvbird) [reply torch/utils/model_dump/code.js:115]: Q. What's Hider? I searched online for "Hider Javascript" but didn't find anything obviously related.
